doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,255 @@
+                    - We want to use the biggest such range in `counts` that doesn’t make `offset` negative. We can do this by an inner `iterate`.
+        - ROWS: similar to GROUPS, but use indexes that include the deduplication component. (see below at “Duplicate indexes”)
+        - There is also `frame_exclusion`, which sometimes necessitates special handling for the group that contains the current row. In such cases, we will put together the result value of the window function from 3 parts: 1. prefix sum (generalized to arbitrary intervals) for groups that are earlier than the current row’s group, 2. prefix sum for groups that are later than the current row’s group, 3. current row’s group (without prefix sum).
+    - Solving 2.:
+        - For invertible aggregation functions (e.g., sum, but not min/max) we can use the existing prefix sum with a minor trick: agg(a,b) = agg(0,b) - agg(0,a).
+            - However, the performance of this might not be good, because even if (a,b) is a small interval, the (0,a) and the (0,b) intervals will be big, so there will be many changes of the aggregates of these even for small input changes.
+        - To have better performance (and to support non-invertible aggregations, e.g., min/max), we need to extend what the `broadcast` part of prefix sum is doing (`aggregate` can stay the same):
+            - `queries` will contain intervals specified by two indexes.
+            - `requests`: We can similarly compute a set of requests from `queries`. The change will only be inside the `flat_map`.
+            - `full_ranges`, `zero_ranges`, `used_ranges` stay the same.
+            - `init_states` won’t start at position 0, but at the lower end of the intervals in `queries`
+            - The iteration at the end will be mostly the same.
+    - FIRST_VALUE / LAST_VALUE with various frames
+        - Can be similarly implemented to window aggregations, i.e., we could “sum” up the relevant interval (that is not necessarily a prefix) with an appropriate sum function.
+        - Or we can make it a bit faster: we just find the index of the relevant end of the interval (i.e., left end for FIRST_VALUE), and then self-join.
+        - (And there are some special cases which we can more efficiently support: FIRST_VALUE with UNBOUNDED PRECEDING and LAST_VALUE with UNBOUNDED FOLLOWING should be transformed to just a (non-windowed) grouped aggregation + self-join instead of prefix sum trickery. Also, similarly for the case when there is no ORDER BY.)
+
+**Duplicate indexes.** There might be rows which agree on both the PARTITION BY key and the ORDER BY key (the index). Prefix Sum doesn’t handle such duplicate indexes.
+
+- To eliminate these duplicates, we will number the elements inside each group with 0, 1, 2, …, and this will be an additional component of the prefix sum indexes.
```

Comment:
The problem with that is that the entire `Row` might match. Note that in other ordering situations this possibility is not a problem, because if the entire `Row` matches, then there is no difference between ordering them one way or the other. But the issue here is not ordering stability, but that the current Prefix Sum implementation just cannot handle duplicate ordering keys.

I added a note that an alternative would be to make Prefix Sum correctly handle duplicate ordering keys.